src/signer/mod.rs

```diff
@@ -55,84 +48,118 @@ impl KeyManager for CryptoBox {
+            seeds: Box::new(std::iter::empty()),
         })
     }
 
-    pub fn derive_from_seed(seeds: &[&str]) -> Result<Self, Error> {
-        let (pub_key, priv_key) = match seeds.get(0) {
-            Some(secret) => {
-                let secret = SecretKey::from_bytes(secret.as_bytes()).map_err(|_| {
-                    Error::SemanticError("failed to convert provided seet to SecretKey".into())
-                })?;
-                let public = PublicKey::from(&secret);
-                (public, secret)
-            }
-            None => ed_new_public_private(),
-        };
+    pub fn derive_from_seed(seeds: Vec<SeedPrefix>) -> Result<Self, Error> {
+        let mut seeds = seeds.into_iter();
+        let (pub_key, priv_key) = derive_key_pair(seeds.next())?;
 
-        let (next_pub_key, next_priv_key) = match seeds.get(1) {
-            Some(secret) => {
-                let seed: SeedPrefix = secret.parse()?;
-                seed.derive_key_pair()?
-            }
-            None => {
-                let (vk, sk) = ed_new_public_private();
-                let vk = Key::new(vk.to_bytes().to_vec());
-                let sk = Key::new(sk.to_bytes().to_vec());
-                (vk, sk)
-            }
-        };
-
-        let seeds = seeds
-            .get(2..)
-            .unwrap_or(&vec![])
-            .iter()
-            .map(|s| s.to_string())
-            .collect();
+        let (next_pub_key, next_priv_key) = derive_key_pair(seeds.next())?;
 
         Ok(CryptoBox {
             signer: Signer {
-                pub_key: Key::new(pub_key.to_bytes().to_vec()),
-                priv_key: Key::new(priv_key.to_bytes().to_vec()),
+                pub_key: pub_key,
+                priv_key: priv_key,
             },
             next_pub_key,
             next_priv_key,
-            seeds,
+            seeds: Box::new(seeds),
         })
     }
 }
 
 struct Signer {
-    priv_key: Key,
-    pub pub_key: Key,
+    priv_key: PrivateKey,
+    pub pub_key: PublicKey,
 }
 
 impl Signer {
-    pub fn new() -> Result<Self, Error> {
+    pub fn new() -> Self {
         let ed = Keypair::generate(&mut OsRng);
-        let pub_key = Key::new(ed.public.to_bytes().to_vec());
-        let priv_key = Key::new(ed.secret.to_bytes().to_vec());
+        let pub_key = PublicKey::new(ed.public.to_bytes().to_vec());
+        let priv_key = PrivateKey::new(ed.secret.to_bytes().to_vec());
 
-        Ok(Signer { pub_key, priv_key })
+        Signer { pub_key, priv_key }
     }
 
-    pub fn sign(&self, msg: &Vec<u8>) -> Result<Vec<u8>, Error> {
+    pub fn sign(&self, msg: &[u8]) -> Result<Vec<u8>, Error> {
         self.priv_key.sign_ed(msg)
     }
 }
 
-fn ed_new_public_private() -> (PublicKey, SecretKey) {
+fn ed_new_public_private() -> (ed25519_dalek::PublicKey, SecretKey) {
     let kp = Keypair::generate(&mut OsRng {});
     (kp.public, kp.secret)
 }
+
+fn derive_key_pair(seed: Option<SeedPrefix>) -> Result<(PublicKey, PrivateKey), Error> {
```

Comment:
this is not a derivation, but new key generation. should'nt we just take single seed in `derive_from_seed()` method and avoid all of these shenanigans with iter it to get first element ONLY, check if it's Some and so forth?

Reply:
The intention of Seeds iterator was to set keypairs that will be used in the next rotations in a way, that enables implementing `KeyManager`'s `rotate` method with no arguments. If we get single seed in `derive_from_seed` argument we can't control next keypairs unless we provide some other mechanism for that. Probably the `derive_from_seed` is not the best name here.

Reply:
in that case maybe add second Option'al parameter of next_seed for that to match exact meaning instead of iterator here?